### PR TITLE
[semantics] avoid sync* and excessive list copies

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Previous benchmarks have established the performance of sync*. Additionally, why allocate a list when you don't need to? Since these APIs are private, they can be re-arranged a bit to avoid creating as many lists.

Will probably not have a large effect on benchmarks

### CODE REVIEW

File: packages/flutter/lib/src/rendering/object.dart [hunk 1]
```diff
@@ -2689,7 +2692,9 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
           continue;
         if (!config.isCompatibleWith(fragment.config))
           toBeMarkedExplicit.add(fragment);
-        for (final _InterestingSemanticsFragment siblingFragment in fragments.sublist(0, fragments.length - 1)) {
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/rendering/object.dart:2692]: avoid creating new list with sublist(0)

File: packages/flutter/lib/src/rendering/object.dart [hunk 2]
```diff
@@ -3577,16 +3581,16 @@ class _RootSemanticsFragment extends _InterestingSemanticsFragment {
...
-        assert(fragment.config == null);
-        return fragment.compileChildren(
-          parentSemanticsClipRect: parentSemanticsClipRect,
-          parentPaintClipRect: parentPaintClipRect,
-          elevationAdjustment: 0.0,
-        );
-      })
-      .toList();
+    final List<SemanticsNode> children = <SemanticsNode>[];
+    for (final _InterestingSemanticsFragment fragment in _children) {
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/rendering/object.dart:3585]: avoid creating more than a single list or any extra iterables

File: packages/flutter/lib/src/rendering/object.dart [hunk 3]
```diff
@@ -3650,19 +3654,20 @@ class _SwitchableSemanticsFragment extends _InterestingSemanticsFragment {
...
   @override
-  Iterable<SemanticsNode> compileChildren({ Rect? parentSemanticsClipRect, Rect? parentPaintClipRect, required double elevationAdjustment }) sync* {
+  void compileChildren({ Rect? parentSemanticsClipRect, Rect? parentPaintClipRect, required double elevationAdjustment, required List<SemanticsNode> result }) {
     if (!_isExplicit) {
       owner._semantics = null;
       for (final _InterestingSemanticsFragment fragment in _children) {
         assert(_ancestorChain.first == fragment._ancestorChain.last);
-        fragment._ancestorChain.addAll(_ancestorChain.sublist(1));
-        yield* fragment.compileChildren(
+        fragment._ancestorChain.addAll(_ancestorChain.skip(1));
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/rendering/object.dart:3662]: avoid creating extra list

File: packages/flutter/lib/src/rendering/object.dart [hunk 4]
```diff
@@ -3699,21 +3704,21 @@ class _SwitchableSemanticsFragment extends _InterestingSemanticsFragment {
       }
     }
 
-    final List<SemanticsNode> children = _children
-      .expand((_InterestingSemanticsFragment fragment) => fragment.compileChildren(
+    final List<SemanticsNode> children = <SemanticsNode>[];
+    for (final _InterestingSemanticsFragment fragment in _children) {
+      fragment.compileChildren(
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/rendering/object.dart:3709]: avoid creating extra iterables

**Comment by jonahwilliams** [comment]
The goal of this PR was to not change any behavior, so as long as the existing tests pass :)

File: packages/flutter/lib/src/rendering/object.dart [hunk 5]
```diff
@@ -2689,7 +2692,9 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
           continue;
         if (!config.isCompatibleWith(fragment.config))
           toBeMarkedExplicit.add(fragment);
-        for (final _InterestingSemanticsFragment siblingFragment in fragments.sublist(0, fragments.length - 1)) {
+        final int initialLength = fragments.length - 1;
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/object.dart:2695]: The name is a bit confusing, maybe siblingLength?

File: packages/flutter/lib/src/rendering/object.dart [hunk 6]
```diff
@@ -3424,7 +3429,7 @@ abstract class _SemanticsFragment {
 
   /// Returns [_InterestingSemanticsFragment] describing the actual semantic
   /// information that this fragment wants to add to the parent.
-  Iterable<_InterestingSemanticsFragment> get interestingFragments;
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/object.dart:3427]: is there necessary?

**Review by chunhtai — APPROVED** [review]
LGTM, just some small nit

File: packages/flutter/lib/src/rendering/object.dart (hunk 6 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/rendering/object.dart:3427]: A more specific type is generally better, in this case it is a private API so there should be no concerns.

File: packages/flutter/lib/src/rendering/object.dart (hunk 5 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/rendering/object.dart]: done

**Pull request merged**